kvstore: Replace gocheck with built-in go test

sayboras opened the pull request:
Please refer to individual commits for more details, this PR is focusing on files owned by cilium/kvstore team.

Relates: https://github.com/cilium/cilium/issues/28596

sayboras committed:
pkg/allocator: Replace gocheck with built-in go test

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
pkg/idpool: Replace gocheck with built-in go test

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
pkg/kvstore: Replace gocheck with built-in go test

This is also to remove the un-used MaintenanceMocker struct

Co-authored-by: Marco Iorio <marco.iorio@isovalent.com>
Signed-off-by: Tam Mach <tam.mach@cilium.io>

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -286,58 +277,58 @@ func testAllocator(c *C, maxID idpool.ID, allocatorName string, suffix string) {
-		c.Assert(firstUse, Equals, true)
+		require.Nil(t, err)
+		require.NotEqual(t, 0, id)
+		require.Equal(t, true, new)
```
Nit. Could be replaced with `require.True` (same for all occurrences).

sayboras replied:
thanks and updated.

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -286,58 +277,58 @@ func testAllocator(c *C, maxID idpool.ID, allocatorName string, suffix string) {
-	c.Assert(new, Equals, false)
-	c.Assert(firstUse, Equals, false)
+	require.NotNil(t, err)
+	require.Equal(t, false, new)
```
Nit. Could be replaced with `require.False` (same for all occurrences).

sayboras replied:
Thanks and updated

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -229,55 +220,55 @@ func randomTestName() string {
 }
 
-func testAllocator(c *C, maxID idpool.ID, allocatorName string, suffix string) {
+func testAllocator(t *testing.T, maxID idpool.ID, allocatorName string, suffix string) {
```
Nit. While you are here, the last two parameters seem unused, and could be removed.

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -286,58 +277,58 @@ func testAllocator(c *C, maxID idpool.ID, allocatorName string, suffix string) {
-		c.Assert(id, Not(Equals), 0)
-		c.Assert(new, Equals, true)
-		c.Assert(firstUse, Equals, true)
+		require.Nil(t, err)
```
Nit. Could be replaced with `require.NoError` (same for all occurrences).

sayboras replied:
thanks and updated

giorio94 commented on pkg/allocator/allocator_test.go:
```diff
@@ -365,7 +356,7 @@ func testAllocator(c *C, maxID idpool.ID, allocatorName string, suffix string) {
 	for i := idpool.ID(1); i <= maxID; i++ {
 		key := TestAllocatorKey(fmt.Sprintf("key%04d", i))
-		c.Assert(allocator.localKeys.keys[allocator.encodeKey(key)], IsNil)
+		require.Nil(t, allocator.localKeys.keys[allocator.encodeKey(key)])
```
I'd replace this with `require.NotContains(t, allocator.localKeys.keys, allocator.encodeKey(key))`

sayboras replied:
thanks

giorio94 commented on pkg/idpool/idpool_test.go:
```diff
@@ -74,240 +64,240 @@ func (s *IDPoolTestSuite) TestInsertRemoveIDs(c *C) {
 	// Unique ids must have been leased.
 	sort.Ints(actual)
-	c.Assert(actual, checker.DeepEquals, expected)
+	require.EqualValues(t, actual, expected)
```
`require.ElementsMatch` is more idiomatic (then you can also drop the sort operation above).

sayboras replied:
ah good point :100:

giorio94 commented on pkg/kvstore/base_test.go:
```diff
@@ -40,172 +39,172 @@ func testValue(i int) string {
-	c.Assert(err, IsNil)
-	c.Assert(pairs, HasLen, 0)
+	require.Nil(t, err)
+	require.Equal(t, 0, len(pairs))
```
`require.Len` is more idiomatic (same for all occurrences).

giorio94 commented on pkg/kvstore/kvstore_test.go:
```diff
@@ -6,31 +6,25 @@ package kvstore
-	TestingT(t)
+func setUpIndependentSuite(tb testing.TB) {
+	testutils.IntegrationTest(tb)
 }
```
Nit. I'd drop this function and just call `testutils.IntegrationTest(tb)` directly.

giorio94 commented on pkg/kvstore/etcd_test.go:
```diff
@@ -95,42 +91,44 @@ func (m MaintenanceMocker) MoveLeader(ctx context.Context, transfereeID uint64)
 
 type EtcdHelpersSuite struct{}
 
 var _ = Suite(&EtcdHelpersSuite{})
```
Should be dropped, otherwise the test is not executed.

giorio94 commented on pkg/kvstore/etcd_test.go:
```diff
@@ -272,33 +270,36 @@ endpoints:
-		c.Assert(gotBool, Equals, tt.wantBool, Commentf("Test %d", i))
-		c.Assert(gotSvcName, Equals, tt.wantSvcName, Commentf("Test %d", i))
+		require.Equal(t, tt.wantBool, Commentf("Test %d", i), gotBool)
+		require.Equal(t, tt.wantSvcName, Commentf("Test %d", i), gotSvcName)
```
The `Commentf` part should be replaced as well.

giorio94 commented on pkg/kvstore/base_test.go:
```diff
@@ -6,28 +6,27 @@ package kvstore
 type BaseTests struct{}
 
-func (s *BaseTests) TestLock(c *C) {
+func (bt *BaseTests) testLock(t *testing.T) {
```
BaseTests should be executed for consul as well.

giorio94 reviewed (changes requested):
Thanks @sayboras for taking care of this, very appreciated.

I've left a few suggestions inline about more idiomatic assertion alternatives, which should be pretty easy to address in batch. I'm only requesting changes due to the loss of coverage in the consul test suite, which should also be easy to fix.

The allocator tests seem to be subject to a data race. However, that is not a regression, and I'll have a look into it as a follow-up.

giorio94 commented on pkg/idpool/idpool_test.go:
```diff
@@ -74,240 +64,239 @@ func (s *IDPoolTestSuite) TestInsertRemoveIDs(c *C) {
 
 	sort.Ints(actualIDs)
-	c.Assert(actualIDs, checker.DeepEquals, evenIDs)
+	require.EqualValues(t, evenIDs, actualIDs)
```
Nit. Could be replaced by `require.ElementsMatch`.

sayboras replied:
thanks, I removed the sort above as well.

giorio94 commented on pkg/kvstore/etcd_test.go:
```diff
@@ -14,26 +14,117 @@ import (
+	OnSnapshot    func(ctx context.Context) (io.ReadCloser, error)
+	OnHashKV      func(ctx context.Context, endpoint string, rev int64) (*etcdAPI.HashKVResponse, error)
+	OnMoveLeader  func(ctx context.Context, transfereeID uint64) (*etcdAPI.MoveLeaderResponse, error)
+}
```
The `MaintenanceMocker` changes should be reverted from this commit, as it is now defined twice (and then dropped in the subsequent one).

sayboras replied:
thanks, seems like I have made a mistake with rebase and fixup.

giorio94 reviewed (commented):
Thanks! A couple of additional comments inline, and a patch to address a few issues in the kvstore allocator tests. I still have to look into the rest of the kvstore tests though.

```diff
diff --git a/pkg/kvstore/allocator/allocator_test.go b/pkg/kvstore/allocator/allocator_test.go
index e42a1af0ae95..3e3e1b0d18bf 100644
--- a/pkg/kvstore/allocator/allocator_test.go
+++ b/pkg/kvstore/allocator/allocator_test.go
@@ -12,6 +12,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"k8s.io/apimachinery/pkg/util/rand"
 
@@ -26,47 +27,11 @@ const (
 	testPrefix = "test-prefix"
 )
 
-type AllocatorSuite struct {
-	backend string
-}
-
-func setupAllocatorSuite(tb testing.TB) *AllocatorSuite {
-	testutils.IntegrationTest(tb)
-
-	return &AllocatorSuite{}
-}
-
-type AllocatorEtcdSuite struct {
-	AllocatorSuite
-}
-
-func setupAllocatorEtcdSuite(tb testing.TB) *AllocatorEtcdSuite {
-	testutils.IntegrationTest(tb)
-
-	s := &AllocatorEtcdSuite{
-		AllocatorSuite: *setupAllocatorSuite(tb),
-	}
-	s.backend = "etcd"
-	kvstore.SetupDummy(tb, "etcd")
-
-	return s
-}
-
-type AllocatorConsulSuite struct {
-	AllocatorSuite
-}
-
-func setupAllocatorConsulSuite(tb testing.TB) *AllocatorConsulSuite {
-	testutils.IntegrationTest(tb)
-
-	s := &AllocatorConsulSuite{
-		AllocatorSuite: *setupAllocatorSuite(tb),
-	}
-	s.backend = "consul"
-	kvstore.SetupDummy(tb, "consul")
-
-	return s
-}
+// Configure a generous timeout to prevent flakes when running in a noisy CI environment.
+var (
+	tick    = 10 * time.Millisecond
+	timeout = 5 * time.Second
+)
 
 // FIXME: this should be named better, it implements pkg/allocator.Backend
 type TestAllocatorKey string
@@ -99,17 +64,17 @@ func randomTestName() string {
 	return fmt.Sprintf("%s%s", testPrefix, rand.String(12))
 }
 
-func BenchmarkAllocateEtcd(b *testing.B) {
-	s := setupAllocatorEtcdSuite(b)
-	s.benchmarkAllocate(b)
-}
-
-func BenchmarkAllocateConsul(b *testing.B) {
-	s := setupAllocatorConsulSuite(b)
-	s.benchmarkAllocate(b)
+func BenchmarkAllocate(b *testing.B) {
+	testutils.IntegrationTest(b)
+	for _, backendName := range []string{"etcd", "consul"} {
+		b.Run(backendName, func(b *testing.B) {
+			kvstore.SetupDummyWithConfigOpts(b, backendName, opts(backendName))
+			benchmarkAllocate(b)
+		})
+	}
 }
 
-func (s *AllocatorSuite) benchmarkAllocate(b *testing.B) {
+func benchmarkAllocate(b *testing.B) {
 	allocatorName := randomTestName()
 	maxID := idpool.ID(256 + b.N)
 	backend, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""), kvstore.Client())
@@ -127,23 +92,22 @@ func (s *AllocatorSuite) benchmarkAllocate(b *testing.B) {
 	b.StopTimer()
 }
 
-func BenchmarkRunLocksGCEtcd(b *testing.B) {
-	s := setupAllocatorEtcdSuite(b)
-	s.benchmarkRunLocksGC(b)
-}
-
-func BenchmarkRunLocksGCConsul(b *testing.B) {
-	s := setupAllocatorConsulSuite(b)
-	s.benchmarkRunLocksGC(b)
+func BenchmarkRunLocksGC(b *testing.B) {
+	testutils.IntegrationTest(b)
+	for _, backendName := range []string{"etcd", "consul"} {
+		b.Run(backendName, func(b *testing.B) {
+			kvstore.SetupDummyWithConfigOpts(b, backendName, opts(backendName))
+			benchmarkRunLocksGC(b, backendName)
+		})
+	}
 }
 
-func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
+func benchmarkRunLocksGC(b *testing.B, backendName string) {
 	allocatorName := randomTestName()
 	maxID := idpool.ID(256 + b.N)
-	// FIXME: Did this previousy use allocatorName := randomTestName() ? so TestAllocatorKey(randomeTestName())
+	// FIXME: Did this previously use allocatorName := randomTestName() ? so TestAllocatorKey(randomeTestName())
 	backend1, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""), kvstore.Client())
 	require.NoError(b, err)
-	require.NoError(b, err)
 	allocator, err := allocator.NewAllocator(TestAllocatorKey(""), backend1, allocator.WithMax(maxID), allocator.WithoutGC())
 	require.NoError(b, err)
 	shortKey := TestAllocatorKey("1;")
@@ -151,7 +115,7 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 	staleLocks := map[string]kvstore.Value{}
 	staleLocks, err = allocator.RunLocksGC(context.Background(), staleLocks)
 	require.NoError(b, err)
-	require.Len(b, staleLocks, 0)
+	require.Empty(b, staleLocks)
 
 	var (
 		lock1, lock2 kvstore.KVLocker
@@ -166,10 +130,10 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 		require.NoError(b, err)
 		close(gotLock1)
 		var client kvstore.BackendOperations
-		switch s.backend {
+		switch backendName {
 		case "etcd":
 			client, _ = kvstore.NewClient(context.Background(),
-				s.backend,
+				backendName,
 				map[string]string{
 					kvstore.EtcdAddrOption: kvstore.EtcdDummyAddress(),
 				},
@@ -177,7 +141,7 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 			)
 		case "consul":
 			client, _ = kvstore.NewClient(context.Background(),
-				s.backend,
+				backendName,
 				map[string]string{
 					kvstore.ConsulAddrOption:   kvstore.ConsulDummyAddress(),
 					kvstore.ConsulOptionConfig: kvstore.ConsulDummyConfigFile(),
@@ -191,14 +155,11 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 	}()
 
 	// Wait until lock1 is gotten.
-	require.Nil(b, testutils.WaitUntil(func() bool {
-		select {
-		case <-gotLock1:
-			return true
-		default:
-			return false
-		}
-	}, 5*time.Second))
+	select {
+	case <-gotLock1:
+	case <-time.After(timeout):
+		b.Error("Lock1 not obtained on time")
+	}
 
 	// wait until client2, in line 160, tries to grab the lock.
 	// We can't detect when that actually happen so we have to assume it will
@@ -208,7 +169,7 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 	// Check which locks are stale, it should be lock1 and lock2
 	staleLocks, err = allocator.RunLocksGC(context.Background(), staleLocks)
 	require.NoError(b, err)
-	switch s.backend {
+	switch backendName {
 	case "consul":
 		// Contrary to etcd, consul does not create a lock in the kvstore
 		// if a lock is already being held.
@@ -246,7 +207,7 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 	// GC lock1 because it's the oldest lock being held.
 	staleLocks, err = allocator.RunLocksGC(context.Background(), staleLocks)
 	require.NoError(b, err)
-	switch s.backend {
+	switch backendName {
 	case "consul":
 		// Contrary to etcd, consul does not create a lock in the kvstore
 		// if a lock is already being held. So we have GCed the only lock
@@ -260,14 +221,11 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 
 	// Wait until lock2 is gotten as it should have happen since we have
 	// GC lock1.
-	require.Nil(b, testutils.WaitUntil(func() bool {
-		select {
-		case <-gotLock2:
-			return true
-		default:
-			return false
-		}
-	}, 10*time.Second))
+	select {
+	case <-gotLock2:
+	case <-time.After(timeout):
+		b.Error("Lock2 not obtained on time")
+	}
 
 	// Unlock lock1 because we still hold the local locks.
 	err = lock1.Unlock(context.Background())
@@ -276,20 +234,20 @@ func (s *AllocatorSuite) benchmarkRunLocksGC(b *testing.B) {
 	require.NoError(b, err)
 }
 
-func BenchmarkGCEtcd(b *testing.B) {
-	s := setupAllocatorEtcdSuite(b)
-	s.benchmarkGC(b)
-}
-
-func BenchmarkGCConsul(b *testing.B) {
-	s := setupAllocatorConsulSuite(b)
-	s.benchmarkGC(b)
+func BenchmarkGC(b *testing.B) {
+	testutils.IntegrationTest(b)
+	for _, backendName := range []string{"etcd", "consul"} {
+		b.Run(backendName, func(b *testing.B) {
+			kvstore.SetupDummyWithConfigOpts(b, backendName, opts(backendName))
+			benchmarkGC(b)
+		})
+	}
 }
 
-func (s *AllocatorSuite) benchmarkGC(b *testing.B) {
+func benchmarkGC(b *testing.B) {
 	allocatorName := randomTestName()
 	maxID := idpool.ID(256 + b.N)
-	// FIXME: Did this previousy use allocatorName := randomTestName() ? so TestAllocatorKey(randomeTestName())
+	// FIXME: Did this previously use allocatorName := randomTestName() ? so TestAllocatorKey(randomeTestName())
 	backend, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""), kvstore.Client())
 	require.NoError(b, err)
 	allocator, err := allocator.NewAllocator(TestAllocatorKey(""), backend, allocator.WithMax(maxID), allocator.WithoutGC())
@@ -310,7 +268,8 @@ func (s *AllocatorSuite) benchmarkGC(b *testing.B) {
 	require.NoError(b, err)
 	require.NotEqual(b, 0, longID)
 
-	allocator.Release(context.Background(), shortKey)
+	_, err = allocator.Release(context.Background(), shortKey)
+	require.NoError(b, err)
 
 	rateLimiter := rate.NewLimiter(10*time.Second, 100)
 
@@ -320,34 +279,31 @@ func (s *AllocatorSuite) benchmarkGC(b *testing.B) {
 	require.Len(b, keysToDelete, 1)
 	keysToDelete, _, err = allocator.RunGC(rateLimiter, keysToDelete)
 	require.NoError(b, err)
-	require.Equal(b, keysToDelete, 0)
+	require.Len(b, keysToDelete, 0)
 
 	// wait for cache to be updated via delete notification
-	require.Nil(b, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(b, func(c *assert.CollectT) {
 		key, err := allocator.GetByID(context.TODO(), shortID)
-		if err != nil {
-			b.Error(err)
-			return false
-		}
-		return key == nil
-	}, 5*time.Second))
+		assert.NoError(c, err)
+		assert.Nil(c, key)
+	}, timeout, tick)
 
 	key, err := allocator.GetByID(context.TODO(), shortID)
 	require.NoError(b, err)
 	require.Nil(b, key)
 }
 
-func BenchmarkGCShouldSkipOutOfRangeIdentitiesEtcd(b *testing.B) {
-	s := setupAllocatorEtcdSuite(b)
-	s.benchmarkGCShouldSkipOutOfRangeIdentities(b)
-}
-
-func BenchmarkGCShouldSkipOutOfRangeIdentitiesConsul(b *testing.B) {
-	s := setupAllocatorConsulSuite(b)
-	s.benchmarkGCShouldSkipOutOfRangeIdentities(b)
+func BenchmarkGCShouldSkipOutOfRangeIdentities(b *testing.B) {
+	testutils.IntegrationTest(b)
+	for _, backendName := range []string{"etcd", "consul"} {
+		b.Run(backendName, func(b *testing.B) {
+			kvstore.SetupDummyWithConfigOpts(b, backendName, opts(backendName))
+			benchmarkGCShouldSkipOutOfRangeIdentities(b)
+		})
+	}
 }
 
-func (s *AllocatorSuite) benchmarkGCShouldSkipOutOfRangeIdentities(b *testing.B) {
+func benchmarkGCShouldSkipOutOfRangeIdentities(b *testing.B) {
 	// Allocator1: allocator under test
 	backend, err := NewKVStoreBackend(randomTestName(), "a", TestAllocatorKey(""), kvstore.Client())
 	require.NoError(b, err)
@@ -367,7 +323,8 @@ func (s *AllocatorSuite) benchmarkGCShouldSkipOutOfRangeIdentities(b *testing.B)
 	require.NoError(b, err)
 	require.NotEqual(b, 0, shortID1)
 
-	allocator1.Release(context.Background(), shortKey1)
+	_, err = allocator1.Release(context.Background(), shortKey1)
+	require.NoError(b, err)
 
 	// Alloctor2: with a non-overlapping range compared with allocator1
 	backend2, err := NewKVStoreBackend(randomTestName(), "a", TestAllocatorKey(""), kvstore.Client())
@@ -402,14 +359,11 @@ func (s *AllocatorSuite) benchmarkGCShouldSkipOutOfRangeIdentities(b *testing.B)
 	require.Len(b, keysToDelete, 0)
 
 	// Wait for cache to be updated via delete notification
-	require.Nil(b, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(b, func(c *assert.CollectT) {
 		key, err := allocator1.GetByID(context.TODO(), shortID1)
-		if err != nil {
-			b.Error(err)
-			return false
-		}
-		return key == nil
-	}, 5*time.Second))
+		assert.NoError(c, err)
+		assert.Nil(c, key)
+	}, timeout, tick)
 
 	// The key created with allocator1 should be GC'd
 	key, err := allocator1.GetByID(context.TODO(), shortID1)
@@ -422,7 +376,17 @@ func (s *AllocatorSuite) benchmarkGCShouldSkipOutOfRangeIdentities(b *testing.B)
 	require.NotNil(b, key2)
 }
 
-func testAllocator(t *testing.T, maxID idpool.ID, allocatorName string) {
+func TestAllocateCached(t *testing.T) {
+	testutils.IntegrationTest(t)
+	for _, backendName := range []string{"etcd", "consul"} {
+		t.Run(backendName, func(t *testing.T) {
+			kvstore.SetupDummyWithConfigOpts(t, backendName, opts(backendName))
+			testAllocatorCached(t, idpool.ID(32), randomTestName()) // enable use of local cache
+		})
+	}
+}
+
+func testAllocatorCached(t *testing.T, maxID idpool.ID, allocatorName string) {
 	backend, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""), kvstore.Client())
 	require.NoError(t, err)
 	a, err := allocator.NewAllocator(TestAllocatorKey(""), backend,
@@ -439,8 +403,8 @@ func testAllocator(t *testing.T, maxID idpool.ID, allocatorName string) {
 		id, new, newLocally, err := a.Allocate(context.Background(), key)
 		require.NoError(t, err)
 		require.NotEqual(t, 0, id)
-		require.Equal(t, true, new)
-		require.Equal(t, true, newLocally)
+		require.True(t, new)
+		require.True(t, newLocally)
 	}
 
 	// allocate all IDs again using the same set of keys, refcnt should go to 2
@@ -449,8 +413,8 @@ func testAllocator(t *testing.T, maxID idpool.ID, allocatorName string) {
 		id, new, newLocally, err := a.Allocate(context.Background(), key)
 		require.NoError(t, err)
 		require.NotEqual(t, 0, id)
-		require.Equal(t, false, new)
-		require.Equal(t, false, newLocally)
+		require.False(t, new)
+		require.False(t, newLocally)
 	}
 
 	// Create a 2nd allocator, refill it
@@ -467,15 +431,16 @@ func testAllocator(t *testing.T, maxID idpool.ID, allocatorName string) {
 		id, new, newLocally, err := a2.Allocate(context.Background(), key)
 		require.NoError(t, err)
 		require.NotEqual(t, 0, id)
-		require.Equal(t, false, new)
-		require.Equal(t, true, newLocally)
+		require.False(t, new)
+		require.True(t, newLocally)
 
 		a2.Release(context.Background(), key)
 	}
 
 	// release 2nd reference of all IDs
 	for i := idpool.ID(1); i <= maxID; i++ {
-		a.Release(context.Background(), TestAllocatorKey(fmt.Sprintf("key%04d", i)))
+		_, err := a.Release(context.Background(), TestAllocatorKey(fmt.Sprintf("key%04d", i)))
+		require.NoError(t, err)
 	}
 
 	staleKeysPreviousRound := map[string]uint64{}
@@ -490,7 +455,8 @@ func testAllocator(t *testing.T, maxID idpool.ID, allocatorName string) {
 
 	// release final reference of all IDs
 	for i := idpool.ID(1); i <= maxID; i++ {
-		a.Release(context.Background(), TestAllocatorKey(fmt.Sprintf("key%04d", i)))
+		_, err := a.Release(context.Background(), TestAllocatorKey(fmt.Sprintf("key%04d", i)))
+		require.NoError(t, err)
 	}
 
 	// running the GC should evict all entries
@@ -508,17 +474,17 @@ func testAllocator(t *testing.T, maxID idpool.ID, allocatorName string) {
 	a2.Delete()
 }
 
-func TestAllocateCachedEtcd(t *testing.T) {
-	setupAllocatorEtcdSuite(t)
-	testAllocator(t, idpool.ID(32), randomTestName()) // enable use of local cache
-}
-
-func TestAllocateCachedConsul(t *testing.T) {
-	setupAllocatorConsulSuite(t)
-	testAllocator(t, idpool.ID(32), randomTestName()) // enable use of local cache
+func TestKeyToID(t *testing.T) {
+	testutils.IntegrationTest(t)
+	for _, backendName := range []string{"etcd", "consul"} {
+		t.Run(backendName, func(t *testing.T) {
+			kvstore.SetupDummyWithConfigOpts(t, backendName, opts(backendName))
+			testKeyToID(t)
+		})
+	}
 }
 
-func (s *AllocatorSuite) TestKeyToID(t *testing.T) {
+func testKeyToID(t *testing.T) {
 	allocatorName := randomTestName()
 	backend, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""), kvstore.Client())
 	require.NoError(t, err)
@@ -543,7 +509,17 @@ func (s *AllocatorSuite) TestKeyToID(t *testing.T) {
 	require.Equal(t, idpool.ID(10), id)
 }
 
-func testGetNoCache(t *testing.T, maxID idpool.ID, suffix string) {
+func TestGetNoCache(t *testing.T) {
+	testutils.IntegrationTest(t)
+	for _, backendName := range []string{"etcd", "consul"} {
+		t.Run(backendName, func(t *testing.T) {
+			kvstore.SetupDummyWithConfigOpts(t, backendName, opts(backendName))
+			testGetNoCache(t, idpool.ID(256))
+		})
+	}
+}
+
+func testGetNoCache(t *testing.T, maxID idpool.ID) {
 	allocatorName := randomTestName()
 	backend, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""), kvstore.Client())
 	require.NoError(t, err)
@@ -560,8 +536,8 @@ func testGetNoCache(t *testing.T, maxID idpool.ID, suffix string) {
 	longID, new, newLocally, err := allocator.Allocate(context.Background(), key)
 	require.NoError(t, err)
 	require.NotEqual(t, 0, longID)
-	require.Equal(t, true, new)
-	require.Equal(t, true, newLocally)
+	require.True(t, new)
+	require.True(t, newLocally)
 
 	observedID, err := allocator.GetNoCache(context.Background(), key)
 	require.NoError(t, err)
@@ -576,8 +552,8 @@ func testGetNoCache(t *testing.T, maxID idpool.ID, suffix string) {
 	shortID, new, newLocally, err := allocator.Allocate(context.Background(), shortKey)
 	require.NoError(t, err)
 	require.NotEqual(t, 0, shortID)
-	require.Equal(t, true, new)
-	require.Equal(t, true, newLocally)
+	require.True(t, new)
+	require.True(t, newLocally)
 
 	observedID, err = allocator.GetNoCache(context.Background(), shortKey)
 	require.NoError(t, err)
@@ -621,17 +597,17 @@ func TestPrefixMatchesKey(t *testing.T) {
 	}
 }
 
-func TestGetNoCacheEtcd(t *testing.T) {
-	setupAllocatorEtcdSuite(t)
-	testGetNoCache(t, idpool.ID(256), "a") // enable use of local cache
-}
-
-func TestGetNoCacheConsul(t *testing.T) {
-	setupAllocatorConsulSuite(t)
-	testGetNoCache(t, idpool.ID(256), "a") // enable use of local cache
+func TestRemoteCache(t *testing.T) {
+	testutils.IntegrationTest(t)
+	for _, backendName := range []string{"etcd", "consul"} {
+		t.Run(backendName, func(t *testing.T) {
+			kvstore.SetupDummyWithConfigOpts(t, backendName, opts(backendName))
+			testRemoteCache(t)
+		})
+	}
 }
 
-func (s *AllocatorSuite) TestRemoteCache(t *testing.T) {
+func testRemoteCache(t *testing.T) {
 	testName := randomTestName()
 	backend, err := NewKVStoreBackend(testName, "a", TestAllocatorKey(""), kvstore.Client())
 	require.NoError(t, err)
@@ -655,13 +631,13 @@ func (s *AllocatorSuite) TestRemoteCache(t *testing.T) {
 	}
 
 	// wait for main cache to be populated
-	require.Nil(t, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		cacheLen := 0
 		a.ForeachCache(func(id idpool.ID, val allocator.AllocatorKey) {
 			cacheLen++
 		})
-		return cacheLen == 4
-	}, 5*time.Second))
+		assert.EqualValues(c, 4, cacheLen)
+	}, timeout, tick)
 
 	// count identical allocations returned
 	cache := map[idpool.ID]int{}
@@ -700,14 +676,14 @@ func (s *AllocatorSuite) TestRemoteCache(t *testing.T) {
 	}()
 
 	// wait for remote cache to be populated
-	require.Nil(t, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		cacheLen := 0
 		a.ForeachCache(func(id idpool.ID, val allocator.AllocatorKey) {
 			cacheLen++
 		})
 		// 4 local + 4 remote
-		return cacheLen == 8
-	}, 5*time.Second))
+		assert.EqualValues(c, 8, cacheLen)
+	}, timeout, tick)
 
 	// count the allocations in the main cache *AND* the remote cache
 	cache = map[idpool.ID]int{}
@@ -722,3 +698,12 @@ func (s *AllocatorSuite) TestRemoteCache(t *testing.T) {
 		require.Equal(t, 2, cache[i])
 	}
 }
+
+func opts(backendName string) map[string]string {
+	if backendName == "etcd" {
+		// Explicitly set higher QPS than the default to speedup the test
+		return map[string]string{kvstore.EtcdRateLimitOption: "100"}
+	}
+
+	return nil
+}

```

sayboras committed:
allocator: fix data races reported by go test -race

Address two separate data races caused by parallel access of a shared
variable by two go routines, and reported when running go test -race.
Both of them concerned test logic only, and didn't affect production
code at all.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>
Signed-off-by: Tam Mach <tam.mach@cilium.io>

giorio94 reviewed (approved):
Thanks!
